### Glossário ANP
https://www.gov.br/anp/pt-br/acesso-a-informacao/glossario

"C:\Users\GG88\Desktop\Gabriel_testes\Glossários_em_CSV\glossario_anp_processado.csv"

### Glossário Petrobrás
https://comunicabaciadesantos.petrobras.com.br/glossario

"C:\Users\GG88\Desktop\Gabriel_testes\Glossários_em_CSV\glossario_petrobras_processado.csv"

### A PetroReconcavo
https://ri.petroreconcavo.com.br/glossario
Glossário - PetroReconcavo | Relações com Investidores

"C:\Users\GG88\Desktop\Gabriel_testes\Glossários_em_CSV\glossario_petroreconcavo_processado.csv"

### Pré-Sal Petróleo lança Biblioteca Interativa
https://www.presalpetroleo.gov.br/glossario/glossario-a/

"C:\Users\GG88\Desktop\Gabriel_testes\Glossários_em_CSV\glossario_presalpetroleo_processado.csv"

#### GLOSSÁRIO DO ANUÁRIO ESTATÍSTICO BRASILEIRO DO PETRÓLEO, GÁS NATURAL E BIOCOMBUSTÍVEIS
https://www.gov.br/anp/pt-br/centrais-de-conteudo/publicacoes/anuario-estatistico/arquivos-anuario-estatistico-2022/outras-pecas-documentais/glossario-anuario-2022.pdf

"C:\Users\GG88\Desktop\Gabriel_testes\Glossários_em_CSV\glossario_anuario_petroleo.csv"





# tarefa

1 Ler vários CSVs, cada um contendo termo e definição

2 Acrescentar automaticamente a coluna “glossario” com o nome correspondente

3 Unificar todos em um único CSV final

In [ ]:
# CAPTURA DOS DADOS
from google.colab import files

uploaded = files.upload()


Saving glossario_anp_processado.csv to glossario_anp_processado.csv
Saving glossario_anuario_petroleo.csv to glossario_anuario_petroleo.csv
Saving glossario_petrobras_processado.csv to glossario_petrobras_processado.csv
Saving glossario_petroreconcavo_processado.csv to glossario_petroreconcavo_processado.csv
Saving glossario_presalpetroleo_processado.csv to glossario_presalpetroleo_processado.csv


In [ ]:
# consolidação dos dados

import pandas as pd

# Arquivos, referência, separador e linhas a pular (para ignorar URL)
arquivos = [
    ("glossario_anp_processado.csv", "glossario_anp", ',', 0),
    ("glossario_anuario_petroleo.csv", "glossario_anuario_petroleo", ';', 0),
    ("glossario_petrobras_processado.csv", "glossario_petrobras", ',', 1),
    ("glossario_petroreconcavo_processado.csv", "glossario_petroreconcavo", ',', 1),
    ("glossario_presalpetroleo_processado.csv", "glossario_presalpetroleo", ',', 0)
]

dfs = []

for nome_arquivo, referencia, sep, skiprows in arquivos:
    print(f"Lendo arquivo: {nome_arquivo} (sep='{sep}', skiprows={skiprows})")
    try:
        df = pd.read_csv(nome_arquivo, sep=sep, skiprows=skiprows, encoding='utf-8')
    except Exception as e:
        print(f"Erro ao ler {nome_arquivo}: {e}")
        continue

    # Normaliza nomes das colunas: tira espaços, põe minúsculas e remove ':'
    df.columns = df.columns.str.strip().str.lower().str.replace(':', '')

    # Identifica colunas termo e definição (busca parcial)
    col_termo = next((c for c in df.columns if 'termo' in c), None)
    col_definicao = next((c for c in df.columns if 'defin' in c), None)

    if col_termo is None or col_definicao is None:
        print(f"Colunas 'termo' e 'definição' não encontradas no arquivo {nome_arquivo}")
        continue

    # Seleciona e renomeia colunas
    df = df[[col_termo, col_definicao]].copy()
    df.columns = ['TERMO', 'DEFINIÇÃO']

    # Remove linhas com termo ou definição vazios
    df = df.dropna(subset=['TERMO', 'DEFINIÇÃO'])

    # Adiciona coluna de referência
    df['REFERENCIA'] = referencia

    dfs.append(df)

if dfs:
    df_final = pd.concat(dfs, ignore_index=True)
    df_final.to_csv('glossario_consolidado.csv', index=False, sep=';', encoding='utf-8-sig')
    print("CSV consolidado criado com sucesso!")
else:
    print("Nenhum dado foi carregado.")


Lendo arquivo: glossario_anp_processado.csv (sep=',', skiprows=0)
Lendo arquivo: glossario_anuario_petroleo.csv (sep=';', skiprows=0)
Lendo arquivo: glossario_petrobras_processado.csv (sep=',', skiprows=1)
Colunas 'termo' e 'definição' não encontradas no arquivo glossario_petrobras_processado.csv
Lendo arquivo: glossario_petroreconcavo_processado.csv (sep=',', skiprows=1)
Colunas 'termo' e 'definição' não encontradas no arquivo glossario_petroreconcavo_processado.csv
Lendo arquivo: glossario_presalpetroleo_processado.csv (sep=',', skiprows=0)
CSV consolidado criado com sucesso!


In [ ]:
 # unificar pdfs em um únco glossário

from google.colab import files
files.download('glossario_consolidado.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
import pandas as pd

df = pd.read_csv('glossario_consolidado.csv', sep=';', encoding='utf-8-sig')
df.head(10)


,TERMO,DEFINIÇÃO,REFERENCIA
0,Abandono de Campo,"Processo que compreende abandono de poços, des...",glossario_anp
1,Abandono de Poço,Série de operações destinadas a restaurar o is...,glossario_anp
2,Abastecimento Nacional de Combustíveis,"Considerado de utilidade pública, abrange as s...",glossario_anp
3,Acidente,Qualquer evento inesperado que cause danos ao ...,glossario_anp
4,Adequação ao Uso,Condições necessárias para que uma Instalação ...,glossario_anp
5,Ver Álcool Etílico Hidratado Combustível,"Agência Nacional do Petróleo, Gás Natural e Bi...",glossario_anp
6,Água de Injeção,"Água injetada em reservatório, com o objetivo ...",glossario_anp
7,Aguarrás,Produto obtido pelo processo de destilação atm...,glossario_anp
8,Alienação de Bens,"Ato de transferir a terceiros, por quaisquer m...",glossario_anp
9,Amostra Composta,"Amostra representativa do produto, preparada s...",glossario_anp


In [ ]:
import pandas as pd

df = pd.read_csv('glossario_consolidado.csv', sep=';', encoding='utf-8-sig')

# Exibe as primeiras 10 linhas com alinhamento à direita
df.head(10).style.set_properties(**{'text-align': 'right'})


,TERMO,DEFINIÇÃO,REFERENCIA
0,Abandono de Campo,"Processo que compreende abandono de poços, desativação e alienação ou reversão de todas as instalações de produção. RESOLUÇÃO ANP Nº 27, DE 18/10/2006",glossario_anp
1,Abandono de Poço,"Série de operações destinadas a restaurar o isolamento entre os diferentes intervalos permeáveis podendo ser permanente, quando não houver interesse de retorno ao poço; ou temporário, quando por qualquer razão houver interesse de retorno ao poço. RESOLUÇÃO ANP Nº 27, DE 18/10/2006",glossario_anp
2,Abastecimento Nacional de Combustíveis,"Considerado de utilidade pública, abrange as seguintes atividades: I – produção, importação, exportação, refino, beneficiamento, tratamento, processamento, transporte, transferência, armazenagem, estocagem, distribuição, revenda, comercialização, avaliação de conformidade e certificação do petróleo, gás natural e seus derivados; II – produção, importação, exportação, armazenagem, estocagem, distribuição, revenda, comercialização, avaliação de conformidade e certificação do biodiesel; III – comercialização, distribuição, revenda e controle de qualidade de álcool etílico combustível.",glossario_anp
3,Acidente,"Qualquer evento inesperado que cause danos ao meio ambiente ou à saúde humana, prejuízos materiais ao patrimônio próprio ou de terceiros, ocorrência de fatalidades ou ferimentos graves para o pessoal próprio ou para terceiros ou a interrupção das operações da Instalação por mais de 24 (vinte e quatro) horas. RESOLUÇÃO ANP Nº 43, DE 6/12/2007 (vide Regulamento Técnico do SGSO)",glossario_anp
4,Adequação ao Uso,"Condições necessárias para que uma Instalação (ou equipamento) seja projetada, mantida, inspecionada, testada e operada de maneira apropriada para o requerido uso, desempenho, disponibilidade e efetividade. RESOLUÇÃO ANP Nº 43, DE 6/12/2007 (vide Regulamento Técnico do SGSO)",glossario_anp
5,Ver Álcool Etílico Hidratado Combustível,"Agência Nacional do Petróleo, Gás Natural e Biocombustíveis (ANP) A Agência Nacional do Petróleo, Gás Natural e Biocombustíveis (ANP) foi criada pela Lei nº 9.478, de 6 de agosto de 1997. Autarquia especial vinculada ao Ministério de Minas e Energia, tem como atribuições promover a regulação, a contratação e a fiscalização das atividades econômicas integrantes da indústria do petróleo, do gás natural e dos biocombustíveis. LEI Nº 9.478, DE 6/8/1997, e LEI Nº 11.097, DE 13/1/2005",glossario_anp
6,Água de Injeção,"Água injetada em reservatório, com o objetivo de forçar a saída do petróleo da rocha-reservatório, deslocando-o para um poço produtor. Este método é conhecido como “recuperação secundária”, e é empregado quando a pressão do poço torna-se insuficiente para expulsar naturalmente o petróleo.",glossario_anp
7,Aguarrás,"Produto obtido pelo processo de destilação atmosférica de petróleo, com intervalo de temperatura típica (150ºC-210ºC), classificado numa faixa de destilação intermediária entre a nafta pesada e o querosene. Utilizado como solvente e na fabricação de ceras, graxas e tintas.",glossario_anp
8,Alienação de Bens,"Ato de transferir a terceiros, por quaisquer meios, um bem de propriedade do concessionário que teve como propósito original a exploração de petróleo ou gás natural. RESOLUÇÃO ANP Nº 13, DE 23/2/2011",glossario_anp
9,Amostra Composta,"Amostra representativa do produto, preparada segundo os percentuais nos quais o produto encontra-se distribuído nos tanques. PORTARIA ANP Nº 311, DE 27/12/2001",glossario_anp


In [ ]:


import pandas as pd
from IPython.display import display
import ipywidgets as widgets

# Carrega o DataFrame
df = pd.read_csv('glossario_consolidado.csv', sep=';', encoding='utf-8-sig')

# Widgets de texto para filtro
filtro_termo = widgets.Text(
    value='',
    placeholder='Filtrar termo',
    description='Termo:',
    disabled=False
)

filtro_definicao = widgets.Text(
    value='',
    placeholder='Filtrar definição',
    description='Definição:',
    disabled=False
)

# Função para atualizar a visualização filtrada
def filtrar_df(change):
    termo_valor = filtro_termo.value.lower()
    definicao_valor = filtro_definicao.value.lower()

    df_filtrado = df

    if termo_valor:
        df_filtrado = df_filtrado[df_filtrado['TERMO'].str.lower().str.contains(termo_valor)]
    if definicao_valor:
        df_filtrado = df_filtrado[df_filtrado['DEFINIÇÃO'].str.lower().str.contains(definicao_valor)]

    display(df_filtrado[['TERMO', 'DEFINIÇÃO', 'REFERENCIA']].reset_index(drop=True))

# Widget interativo
out = widgets.Output()

def on_text_change(change):
    out.clear_output()
    with out:
        filtrar_df(change)

filtro_termo.observe(on_text_change, names='value')
filtro_definicao.observe(on_text_change, names='value')

display(filtro_termo, filtro_definicao, out)

# Exibe inicialmente tudo
with out:
    display(df[['TERMO', 'DEFINIÇÃO', 'REFERENCIA']].head(20))



Text(value='', description='Termo:', placeholder='Filtrar termo')

Text(value='', description='Definição:', placeholder='Filtrar definição')

Output()

# Processamento de Linguagem Natural da Coluna Definição

In [ ]:
#1. Preparar ambiente
#Se ainda não instalou, instale bibliotecas básicas para PLN:
!pip install -q nltk spacy
!python -m spacy download pt_core_news_sm

#obs "-m" execute módulo como script, usando o Python diretamente, sem precisar chamar um arquivo .py específico.
#Ou, para rodar o comando spacy download diretamente: !spacy download pt_core_news_sm


  Using cached https://github.com/explosion/spacy-models/releases/download/pt_core_news_sm-3.8.0/pt_core_news_sm-3.8.0-py3-none-any.whl (13.0 MB)
✔ Download and installation successful
You can now load the package via spacy.load('pt_core_news_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [ ]:
#2. Carregar dados
import pandas as pd

df = pd.read_csv('glossario_consolidado.csv', sep=';', encoding='utf-8-sig')
df

,TERMO,DEFINIÇÃO,REFERENCIA
0,Abandono de Campo,"Processo que compreende abandono de poços, des...",glossario_anp
1,Abandono de Poço,Série de operações destinadas a restaurar o is...,glossario_anp
2,Abastecimento Nacional de Combustíveis,"Considerado de utilidade pública, abrange as s...",glossario_anp
3,Acidente,Qualquer evento inesperado que cause danos ao ...,glossario_anp
4,Adequação ao Uso,Condições necessárias para que uma Instalação ...,glossario_anp
...,...,...,...
781,Xisto,"xisto betuminoso é uma rocha sedimentar, norma...",glossario_anuario_petroleo
782,Acondicionamento de Gás Natural: confinamento ...,Acordo de Individualização da Produção: acordo...,glossario_presalpetroleo
783,Concessionário:,– empresa que firmou com a ANP um Contrato de ...,glossario_presalpetroleo
784,Concessionário:,– empresa que firmou com a ANP um Contrato de ...,glossario_presalpetroleo


In [ ]:
#3. Pré-processamento básico das definições
#Exemplo para tokenizar, remover stopwords e lematizar:

import spacy

nlp = spacy.load("pt_core_news_sm")  # modelo em português

def preprocessar(texto):
    doc = nlp(str(texto).lower())
    tokens = [token.lemma_ for token in doc if not token.is_stop and not token.is_punct]
    return tokens

df['definicao_tokens'] = df['DEFINIÇÃO'].apply(preprocessar)

df

,TERMO,DEFINIÇÃO,REFERENCIA,definicao_tokens
0,Abandono de Campo,"Processo que compreende abandono de poços, des...",glossario_anp,"[processar, compreender, abandono, poço, desat..."
1,Abandono de Poço,Série de operações destinadas a restaurar o is...,glossario_anp,"[série, operação, destinar, restaurar, isolame..."
2,Abastecimento Nacional de Combustíveis,"Considerado de utilidade pública, abrange as s...",glossario_anp,"[considerar, utilidade, público, abranger, seg..."
3,Acidente,Qualquer evento inesperado que cause danos ao ...,glossario_anp,"[evento, inesperar, causar, dano, ambiente, sa..."
4,Adequação ao Uso,Condições necessárias para que uma Instalação ...,glossario_anp,"[condição, necessário, instalação, equipamento..."
...,...,...,...,...
781,Xisto,"xisto betuminoso é uma rocha sedimentar, norma...",glossario_anuario_petroleo,"[xisto, betuminoso, rocha, sedimentar, normalm..."
782,Acondicionamento de Gás Natural: confinamento ...,Acordo de Individualização da Produção: acordo...,glossario_presalpetroleo,"[acordo, individualização, produção, acordo, c..."
783,Concessionário:,– empresa que firmou com a ANP um Contrato de ...,glossario_presalpetroleo,"[empresa, firmar, anp, contrato, concessão, ex..."
784,Concessionário:,– empresa que firmou com a ANP um Contrato de ...,glossario_presalpetroleo,"[empresa, firmar, anp, contrato, concessão, ex..."


# SPACY

| Coluna            | Significa                      | Vem de onde     |
| ----------------- | ------------------------------ | --------------- |
| **textcat_label** | Categoria real / correta       | Seus dados      |
| **textcat_pred**  | Categoria prevista pelo modelo | Algoritmo spaCy |


### 4. Exemplos de aplicações que você pode fazer com as definições processadas:

Análise de frequência: quais palavras aparecem mais nas definições.

Word embeddings: representar definições como vetores para similaridade ou clustering.

Classificação automática: categorizar termos por tipos.

Resumo automático: gerar resumos das definições.

Busca semântica: melhorar a busca considerando sinônimos e contexto.

# 4.1 Análise de frequência: quais palavras aparecem mais nas definições.


In [ ]:
import pandas as pd
from collections import Counter

# Supondo que 'contador' seja o Counter com as palavras e suas frequências
todas_palavras = [palavra for tokens in df['definicao_tokens'] for palavra in tokens]
contador = Counter(todas_palavras)

# Converte o Counter para DataFrame
df_freq = pd.DataFrame(contador.items(), columns=['Palavra', 'Frequência'])

# Ordena pela frequência decrescente
df_freq = df_freq.sort_values(by='Frequência', ascending=False).reset_index(drop=True)

# Exibe a lista completa
print(df_freq.to_string(index=False))


################################################

# Gera o contador
#todas_palavras = [palavra for tokens in df['definicao_tokens'] for palavra in tokens]
#contador = Counter(todas_palavras)

# DataFrame de frequências
#df_freq = pd.DataFrame(contador.items(), columns=['Palavra', 'Frequência'])
#df_freq = df_freq.sort_values(by='Frequência', ascending=False).reset_index(drop=True)

# Salva o CSV
#caminho = "frequencia_palavras.csv"
#df_freq.to_csv(caminho, index=False, encoding="utf-8-sig")

# Abre o arquivo no painel direito (somente para VSCode ou JupyterLab)
# Jupyter: mostra como "Open in Data Viewer" no lado direito
#df_freq






                                                                                                                                                                                                    Palavra  Frequência
                                                                                                                                                                                                         nº         600
                                                                                                                                                                                                        anp         593
                                                                                                                                                                                                        gás         471
                                                                                                                                        

# Destas planilha foi feita uma uma classificação pelas as seguintes definições arbitrárias a partir das 60 palavras chaves

✔️ Classificações únicas encontradas

### jurídico, instituição, produto, matriz, adjetivo, atividade, definição, local, setor, questionar ?


| Índice | Palavra         | Frequência | Classificação |
| -----: | --------------- | ---------: | ------------- |
|      0 | nº              |        600 | jurídico      |
|      1 | anp             |        593 | instituição   |
|      2 | gás             |        471 | produto       |
|      3 | petróleo        |        397 | matriz        |
|      4 | natural         |        388 | adjetivo      |
|      5 | produção        |        303 | atividade     |
|      6 | resolução       |        302 | jurídico      |
|      7 | combustível     |        166 | produto       |
|      8 | lei             |        149 | jurídico      |
|      9 | portaria        |        136 | jurídico      |
|     10 | volume          |        113 | definição     |
|     11 | óleo            |        112 | produto       |
|     12 | unidade         |        109 | local ou de   |
|     13 | poço            |        105 | local         |
|     14 | transporte      |        105 | atividade     |
|     15 | contrato        |        103 | atividade     |
|     16 | atividade       |        102 | definição     |
|     17 | dou             |        100 | jurídico      |
|     20 | instalação      |         88 | atividade     |
|     21 | hidrocarboneto  |         86 | produto       |
|     22 | campo           |         84 | local         |
|     23 | produto         |         82 | definição     |
|     24 | reservatório    |         81 | local         |
|     25 | conjunto        |         80 | definição     |
|     26 | exploração      |         76 | atividade     |
|     27 | condição        |         72 | definição     |
|     29 | concessão       |         69 | atividade     |
|     30 | líquido         |         68 | definição     |
|     31 | destinar        |         68 | atividade     |
|     32 | medição         |         67 | atividade     |
|     33 | produzir        |         60 | atividade     |
|     34 | desenvolvimento |         58 | atividade     |
|     35 | operação        |         58 | atividade     |
|     36 | processo        |         58 | atividade     |
|     40 | concessionário  |         50 | setor         |
|     41 | processamento   |         50 | atividade     |
|     42 | 21/1/2000       |         50 | questionar    |
|     43 | derivar         |         50 | atividade     |
|     44 | portar          |         49 | atividade     |
|     46 | fluido          |         47 | definição     |
|     47 | derivado        |         46 | produto       |
|     48 | diesel          |         46 | produto       |
|     49 | mistura         |         45 | atividade     |
|     50 | fração          |         45 | definição     |
|     51 | autorizar       |         45 | atividade     |
|     52 | capacidade      |         45 | atividade     |
|     53 | 2009            |         45 | questionar    |
|     55 | energia         |         44 | produto       |
|     58 | pressão         |         43 | atividade     |





Obs: Estruturas matriciais são modelos de organização que combinam departamentalização funcional com a divisão por projeto ou produto, resultando em uma dupla subordinação para os colaboradores, que se reportam a um gerente funcional e a um gerente de projeto simultaneamente. Essa estrutura promove a colaboração entre áreas, utiliza recursos de forma eficiente e aumenta a flexibilidade para lidar com projetos complexos


### Características principais

Dupla autoridade: O ponto central da estrutura matricial é que os funcionários têm dois chefes: um gerente funcional (responsável pela área de especialidade, como marketing ou engenharia) e um gerente de projeto (responsável por um projeto específico).

Equipes multidisciplinares: Profissionais de diferentes departamentos são alocados para trabalhar juntos em projetos específicos, compartilhando conhecimentos e habilidades.

Flexibilidade: É ideal para empresas que precisam gerenciar múltiplos projetos ao mesmo tempo, pois permite a rápida realocação de recursos e equipes para diferentes iniciativas.

### Vantagens

Otimização de recursos: Permite o uso mais eficiente de talentos e habilidades especializadas, já que o mesmo profissional pode trabalhar em diferentes projetos.

Melhora da comunicação: A comunicação entre os diferentes setores é fortalecida, pois as equipes se reúnem para trabalhar em objetivos comuns de projeto.

Desenvolvimento de colaboradores: Oferece aos funcionários a oportunidade de trabalhar em projetos variados, o que pode aumentar o engajamento e expandir suas habilidades.

### Desvantagens

Potencial para conflitos: A dupla subordinação pode gerar conflitos entre os gerentes sobre prioridades e recursos, e os funcionários podem ficar confusos sobre a quem devem obedecer.

Necessidade de comunicação clara: Requer uma comunicação constante e clara entre os gerentes e a equipe para garantir o alinhamento e evitar atritos.
Complexidade: É uma estrutura mais complexa do que as tradicionais, exigindo maturidade organizacional e líderes capacitados para gerenciar a dualidade de comando.